klippy/extras/temperature_fan.py

```diff
@@ -133,6 +137,31 @@ def temperature_callback(self, read_time, temp):
             self.temperature_fan.set_speed(read_time,
                                            self.temperature_fan.get_max_speed())
 
+######################################################################
+# Linear control algo
+# controls the fan speed based on a linear temperature range between
+# the fans min and max speeds
+######################################################################
+
+class ControlLinear:
+    def __init__(self, temperature_fan, config):
+        self.temperature_fan = temperature_fan
+        self.linear_offset = config.getfloat('linear_offset', 10, above=0.)
+        self.min_speed = self.temperature_fan.get_min_speed()
+        self.max_speed = self.temperature_fan.get_max_speed()
+    def temperature_callback(self, read_time, temp):
+        current_temp, target_temp = self.temperature_fan.get_temp(read_time)
+        temp_diff = target_temp - current_temp
+        # calculate linear speed range between 0.0 and 1.0
+        speed = 1 - (temp_diff / self.linear_offset)
+        # convert it to respect min/max speed
+        speed = self.min_speed + (speed * (self.max_speed - self.min_speed))
```

Comment:
Test case: T=50, target=45, lin=10, min=0, max=100
speed= 1-(-5/10)= 1.5
speed= 0+1.5*(100-0)= 100 (capped)
So it goes 100% even if the target temp is lower than the current one.
Maybe the sign is wrong.

Reply:
I'm not sure about coding style, but you are reusing the variable name speed for two things which are different. Memory-wise it won't change anything since a new variable has to be allocated in memory, but legibility suffers.
Use "compensation =" for the first "speed =" (line 156)

Reply:
Also, with T = Ttarget the first "speed" is 1, which means the fan goes to full?

I think this should be rewritten along the principle:

speed = current speed + temp_diff/self.linear_offset * (self.max_speed - self.min_speed)

Which means: T = target temp -> no changes.
T > target temp -> speed decreased. How much it depends on the linear_offset.

I notice that this is becoming a proportional setup, with no D component and with a make-up I(ntegral) component, but as you coded, it also makes little sense because it works strange (to me! I may have misunderstood).

Reply:
Mybe I'm wrong but for me it reads like:
`temp_diff = target_temp - current_temp`
Test case: target=45, lin=10, min=0, max=100

At 35°C current temp:
temp_diff = 45-35=10
speed = 1-(10/10) = 0
--> Correct, since we want to start when the temperature is above 35

At 40°C
temp_diff = 45-40=5
speed = 1-(5/10) = 0.5
--> Correct, fan at half power

At 45°C
temp_diff = 45-45=0
speed = 1-(0/10) = 1
--> Correct, fan at full power

Reply:
My mistake, I read it again and I undestood.

But then the explanation is poorly written.
Please change it from "linear_offset" to "linear_starting_temp " and the explanation to something along the lines of:

```
#   In addition to "watermark" and "pid", temperature_fan has an additional
#   "linear" control that sets the fan speed to a value between min_speed
#   and max_speed, starting from linear_starting_temp and the target_temp.
#linear_starting_temperature: 30.0
#   on "linear" control, this sets the temperature at which the linear scaling starts
```

Why making things complicated when they can be straightforward? :)

Reply:
I agree, the term `linear_offset` is not self explanatory. Maybe it would be simpler not to provide a delta but an absolute starting temperature and called it `start_temperature`, `lower_threshold` or something along these lines.

Reply:
Thanks for reviewing!

@dewi-ny-je glad you got it correctly now :)

<img width="382" alt="image" src="https://user-images.githubusercontent.com/769791/161603176-d158308a-101d-4f90-9eff-d2ce0962c650.png">
@Sineos One can change the fans target temp on runtime using "SET_TEMPERATURE_FAN_TARGET TEMPERATURE_FAN=psu TARGET=45" (or using fluidd for example). Having a absolute "linear_starting_temp" would mean that one would be able to change the target temp below the starting_temp. That case then would need to be handled one way or the other and may yield (for the users point of view) unexpected results that may require additional documentation. Changing it closer to the staring temp would mean a more narrow scaling range, which may also be undesired.
I can think of the use-case that someone likes his components to run as cool as possible (to increase lifetime) most of the time, but also wants to be able to make them more quiet (for example on night time) when required. Another case is that one wants to test different target temps during a print to find a good noise-temp trade-off.

As @dewi-ny-je had initial issues understanding what it was meant to do, the documentation may need improvement. As i am no native speaker, i appreciate any help on that. I have initially thought a while about the name, as i was not easy to find something self explanatory that was not relatively long. here are some thoughts:

1. linear_* should be the prefix to make clear that this option only relates to the linear scaling.
2. linear_offset (current)
3. linear_delta (i dont like the word delta, as it sounds a bit too scientific for what this option does)
4. linear_start_offset (a bit more verbose)
5. linear_scaling_range (that one could be better, as it defines over how many degrees the scaling takes place)
6. linear_scaling_amount (not sure on that)

I am open for ideas!

Reply:
I see your point on setting the delta, so maybe it's ok if you use "linear_scaling_range" but also the description I provided, which makes easier to understand what it does.

Concerning your use case, I thought that a fixed starting point made sense because making the fans quieter would mean just increasing the target temperature: why changing the (in my idea, fixed) starting one? at (for example) 35 °C the cpu simply needs no cooling, ever. It makes sense to me, as user, to fix a starting temperature when sor dure I don't need cooling, and change the upper extreme depending on the other conditions. But I guess few people will change it on the fly anyway... so the "linear_scaling_range" is fine to me.